### PR TITLE
add market gas price to simulate duplicate nonce behavior

### PR DESCRIPTION
**Pull request by susannapaxos**
add field of "marketGasPrice" to the simulated backend

this is useful when we want to test some duplicate nonce issue and more align with the actual chain behavior. (with tx below that price, it wont be committed)

note that i tried to use this commit with pax repo: https://github.com/paxosglobal/pax/pull/8122 (draft because i need to change my branch after this merged) and they built successfully together.

### CODE REVIEW

File: accounts/abi/bind/backends/simulated.go [hunk 1]
```diff
@@ -210,6 +212,13 @@ func (b *SimulatedBackend) NonceAt(ctx context.Context, contract common.Address,
 	return stateDB.GetNonce(contract), nil
 }
 
+func (b *SimulatedBackend) SetMarketGasPrice(gasPrice int64) {
```
Comment (gitteri) [inline accounts/abi/bind/backends/simulated.go:233]: let's add a go style comment here describing this function and a potential use case

Comment (susannapaxos) [reply accounts/abi/bind/backends/simulated.go:233]: done!

File: accounts/abi/bind/backends/simulated.go [hunk 2]
```diff
@@ -645,9 +656,13 @@ func (b *SimulatedBackend) SendTransaction(ctx context.Context, tx *types.Transa
 	// Include tx in chain
 	blocks, _ := core.GenerateChain(b.config, block, ethash.NewFaker(), b.database, 1, func(number int, block *core.BlockGen) {
 		for _, tx := range b.pendingBlock.Transactions() {
+			if b.marketGasPrice == nil || b.marketGasPrice.Cmp(tx.GasPrice())  <= 0 {
+				block.AddTxWithChain(b.blockchain, tx)
+			}
+		}
+		if b.marketGasPrice == nil || b.marketGasPrice.Cmp(tx.GasPrice())  <= 0 {
 			block.AddTxWithChain(b.blockchain, tx)
 		}
```
Comment (lsharir) [inline accounts/abi/bind/backends/simulated.go]: could use a small comment explaining what is being checked here
also, there could be scenarios where only some of the pending transactions and sent transactions get added to the block. Will this zero out the remaining "low gas price" pending transactions?

Comment (susannapaxos) [reply accounts/abi/bind/backends/simulated.go]: so i tried to only have changes in the `Commit()` function and keep the `SendTransaction` intact and it seems like if we dont modify the `pendingBlock` we are unable to do duplicate nonce in simulated chain.

anything that is in our pendingBlock even when its not in the chain when we `commit()` it changes the state of the db, any new transaction check this state of the db and requires the nonce has to be greater than whatever is in the pendingBlock.  

i did add a stuck transactions field and have test against that scenario, let me know if that works. Thank you!

File: accounts/abi/bind/backends/simulated.go [hunk 3]
```diff
@@ -109,12 +112,37 @@ func (b *SimulatedBackend) Commit() {
...
+				}
+			}
+
+			for _, tx := range b.stuckTransactions {
+				if b.marketGasPrice == nil || b.marketGasPrice.Cmp(tx.GasPrice()) <= 0 {
+					block.AddTxWithChain(b.blockchain, tx)
+				} else {
+					remainingTx = append(remainingTx, tx)
+				}
+			}
```
Comment (gipsy86147) [inline accounts/abi/bind/backends/simulated.go]: nit: 
```
                        for _, tx := range append(b.pendingBlock.Transactions(), b.stuckTransactions...) {
				if b.marketGasPrice == nil || b.marketGasPrice.Cmp(tx.GasPrice()) <= 0 {
					block.AddTxWithChain(b.blockchain, tx)
				} else {
					remainingTx = append(remainingTx, tx)
				}
			}
```
Can this work with a single loop? Not saving anything just less lines of code

Comment (susannapaxos) [reply accounts/abi/bind/backends/simulated.go]: thanks for the suggestion, its neat!

**Review by vlfurman — APPROVED** [review]

File: accounts/abi/bind/backends/simulated.go [hunk 4]
```diff
@@ -126,8 +145,8 @@ func (b *SimulatedBackend) Rollback() {
 }
 
 func (b *SimulatedBackend) rollback(parent *types.Block) {
-	blocks, _ := core.GenerateChain(b.config, parent, ethash.NewFaker(), b.database, 1, func(int, *core.BlockGen) {})
 
+	blocks, _ := core.GenerateChain(b.config, parent, ethash.NewFaker(), b.database, 1, func(int, *core.BlockGen) {})
```
Comment (gitteri) [inline accounts/abi/bind/backends/simulated.go:148]: nit: why did we add a new line above this?

File: accounts/abi/bind/backends/simulated_test.go [hunk 5]
```diff
@@ -1336,3 +1339,129 @@ func TestForkResendTx(t *testing.T) {
...
+	defer sim.Close()
+
+	bgCtx := context.Background()
+
+	// Create tx and send
+	head, _ := sim.HeaderByNumber(context.Background(), nil) // Should be child's, good enough
+	gasPrice := new(big.Int).Add(head.BaseFee, big.NewInt(1))
+	sim.SetMarketGasPrice(int64(gasPrice.Uint64()))
+
+	lowGasFeeTx := types.NewTx(&types.LegacyTx{
```
Comment (gitteri) [inline accounts/abi/bind/backends/simulated_test.go]: nit: let's try both legacy and dynamic fee txs in this test

**Review by gitteri — APPROVED** [review]

File: accounts/abi/bind/backends/simulated_test.go [hunk 6]
```diff
@@ -1336,3 +1339,129 @@ func TestForkResendTx(t *testing.T) {
...
+
+	// send tx to simulated backend
+	err = sim.SendTransaction(bgCtx, signedTx)
+	require.NoError(t, err)
+	sim.Commit()
+
+	// check that the nonce stays the same because nothing has been mined
+	pendingNonce, err := sim.PendingNonceAt(bgCtx, testAddr)
+	require.NoError(t, err)
+	assert.Equal(t, uint64(0), pendingNonce)
```
Comment (lsharir) [inline accounts/abi/bind/backends/simulated_test.go]: let's do another commit and nonce check after this one? so we go through the cycle of passing forward the stuckTransactions twice.

File: accounts/abi/bind/backends/simulated.go [hunk 7]
```diff
@@ -109,12 +112,28 @@ func (b *SimulatedBackend) Commit() {
...
+			}
+		})
+
+	if _, err := b.blockchain.InsertChain([]*types.Block{blocks[0]}); err != nil {
 		panic(err) // This cannot happen unless the simulator is wrong, fail in that case
 	}
 	// Using the last inserted block here makes it possible to build on a side
 	// chain after a fork.
-	b.rollback(b.pendingBlock)
+	b.stuckTransactions = remainingTx
```
Comment (lsharir) [inline accounts/abi/bind/backends/simulated.go:135]: can we instead just put them back on the pending block?
similarly to how SendTransaction recreates the pendingBlock

Comment (susannapaxos) [reply accounts/abi/bind/backends/simulated.go:135]: we cannot put them to the pending block, the txes in the pending block determine whats the next nonce available. if they are stuck, they should not be in pending block. because that means we cannot send the transaction with the same nonce. it will be invalid/nonce too low.

File: accounts/abi/bind/backends/simulated_test.go (hunk 6 above)
Comment (susannapaxos) [reply accounts/abi/bind/backends/simulated_test.go]: this test set up is the following
(1). Send a tx lower than the set market gas price >> tx ends up in stuck transactions
(2). call commit to chain >> tx stills in stuck transactions
(3). drop gas fee, call commit to chain >> stuck transactions txes get mined.

hmmm, are you suggesting after (3) we should commit again even though we dont have new tx and nothing in stuck transactions?

File: accounts/abi/bind/backends/simulated_test.go (hunk 5 above)
Comment (susannapaxos) [reply accounts/abi/bind/backends/simulated_test.go]: added both legacy and dynamic fee txs in both test cases!

File: accounts/abi/bind/backends/simulated_test.go [hunk 8]
```diff
@@ -1336,3 +1339,195 @@ func TestForkResendTx(t *testing.T) {
...
+		{
+			name: "dynamicTx",
+			txType: types.DynamicFeeTxType,
+		},
+	}
+	for i, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			bgCtx := context.Background()
+			// Create tx and send
+			head, _ := sim.HeaderByNumber(context.Background(), nil) // Should be child's, good enough
```
Comment (gitteri) [inline accounts/abi/bind/backends/simulated_test.go]: you can remove this comment

File: accounts/abi/bind/backends/simulated_test.go [hunk 9]
```diff
@@ -1336,3 +1339,195 @@ func TestForkResendTx(t *testing.T) {
...
+			gasPrice := new(big.Int).Add(head.BaseFee, big.NewInt(1))
```
Comment (gitteri) [inline accounts/abi/bind/backends/simulated_test.go]: nit: can use get suggested fee tip method I believe instead of just hard coding 1

File: accounts/abi/bind/backends/simulated_test.go [hunk 10]
```diff
@@ -1336,3 +1339,195 @@ func TestForkResendTx(t *testing.T) {
...
+
+			lowFeeGas := new(big.Int).Sub(gasPrice, big.NewInt(1))
+
+			var lowGasFeeTx *types.Transaction
+
+			if test.txType == types.LegacyTxType {
+				lowGasFeeTx = types.NewTx(&types.LegacyTx{
+					Nonce:    uint64(i),
+					To:       &testAddr,
+					Value:    big.NewInt(1),
```
Comment (gitteri) [inline accounts/abi/bind/backends/simulated_test.go]: nit: make this a variable

**Review by gitteri — APPROVED** [review]
